--- a/Cargo.toml
+++ b/Cargo.toml
@@ -91,7 +91,7 @@
 version = "0.45"
 
 [target.'cfg(any(target_os = "linux", target_os = "dragonfly", target_os = "freebsd", target_os = "netbsd", target_os = "openbsd"))'.dependencies.gtk]
-version = "0.18"
+version = "0.19"
 optional = true
 
 [target.'cfg(any(target_os = "linux", target_os = "dragonfly", target_os = "freebsd", target_os = "netbsd", target_os = "openbsd"))'.dependencies.libxdo]
